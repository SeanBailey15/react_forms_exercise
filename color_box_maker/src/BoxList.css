.BoxList-h1 {
  text-align: center;
}

.BoxList-form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin: 10px;
}

.BoxList-boxes {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-top: 50px;
}
