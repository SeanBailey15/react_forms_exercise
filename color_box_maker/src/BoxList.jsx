import React, {useState} from "react";
import {v4 as uuid} from 'uuid'
import Box from './Box'
import NewBoxForm from './NewBoxForm'

const BoxList = () => {
    const INITIAL_STATE = []

    const [boxes, setBoxes] = useState(INITIAL_STATE);

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id: uuid()}])
    }

    const removeBox = (id) => {
       setBoxes(boxes.filter((box) => box.id !== id))     
    }

    return (
        <div className="BoxList">
            <h3 className="BoxList-h3">Make Some Boxes!</h3>
            <div className="BoxList-form">
                <NewBoxForm addBox={addBox}/>
            </div>
            <div className="BoxList-boxes">
                {boxes.map(({id, width, height, color}) => <Box key={id} id={id} width={Number(width)} height={Number(height)} color={color} removeBox={removeBox} />)}
            </div>
        </div>
    )
}

export default BoxList;