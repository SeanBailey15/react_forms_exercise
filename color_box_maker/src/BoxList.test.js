import React from "react";
import { render, fireEvent, screen, act } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", () => {
  act(() => {
    render(<BoxList />);
  });
});

it("matches snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a new box", async () => {
  act(() => {
    render(<BoxList />);
  });
  const width = screen.getByLabelText("Box Width");
  const height = screen.getByLabelText("Box Height");
  const color = screen.getByLabelText("Box Color");
  const addButton = screen.queryByText("Add Box");

  //check if button to remove a box is present
  expect(screen.queryByText("X")).not.toBeInTheDocument();

  //input box dimensions and color, then click to submit
  fireEvent.change(width, { target: { value: 100 } });
  fireEvent.change(height, { target: { value: 100 } });
  fireEvent.change(color, { target: { value: "blue" } });
  fireEvent.click(addButton);

  //check that button to remove a box is now present
  const closeButton = await screen.findByText("X");
  expect(closeButton).toBeInTheDocument();

  const box = screen.getByTestId("box");
  expect(box).toBeInTheDocument();
  expect(box).toHaveStyle(
    "width: 100px; height: 100px; background-color: blue;"
  );
});

it("should remove a box", async () => {
  act(() => {
    render(<BoxList />);
  });
  const width = screen.getByLabelText("Box Width");
  const height = screen.getByLabelText("Box Height");
  const color = screen.getByLabelText("Box Color");
  const addButton = screen.queryByText("Add Box");

  expect(screen.queryByText("X")).not.toBeInTheDocument();

  fireEvent.change(width, { target: { value: 100 } });
  fireEvent.change(height, { target: { value: 100 } });
  fireEvent.change(color, { target: { value: "blue" } });
  fireEvent.click(addButton);

  const removeButton = await screen.findByText("X");
  expect(removeButton).toBeInTheDocument();

  const box = screen.getByTestId("box");
  expect(box).toBeInTheDocument();
  expect(box).toHaveStyle(
    "width: 100px; height: 100px; background-color: blue;"
  );

  fireEvent.click(removeButton);

  expect(box).not.toBeInTheDocument();
  expect(removeButton).not.toBeInTheDocument();
});
