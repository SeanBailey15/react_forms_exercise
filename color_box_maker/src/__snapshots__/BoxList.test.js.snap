// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h1
      class="BoxList-h1"
    >
      Make Some Boxes!
    </h1>
    <div
      class="BoxList-form"
    >
      <div
        class="NewBoxForm"
      >
        <form
          class="NewBoxForm-form"
        >
          <label
            class="NewBoxForm-label"
            for="width"
          >
            Box Width
          </label>
          <input
            class="NewBoxForm-input"
            id="width"
            name="width"
            placeholder="Pixels"
            type="text"
            value=""
          />
          <label
            class="NewBoxForm-label"
            for="height"
          >
            Box Height
          </label>
          <input
            class="NewBoxForm-input"
            id="height"
            name="height"
            placeholder="Pixels"
            type="text"
            value=""
          />
          <label
            class="NewBoxForm-label"
            for="color"
          >
            Box Color
          </label>
          <input
            class="NewBoxForm-input"
            id="color"
            name="color"
            placeholder="Color Name"
            type="text"
            value=""
          />
          <button
            class="NewBoxForm-btn"
            type="submit"
          >
            Add Box
          </button>
        </form>
      </div>
    </div>
    <div
      class="BoxList-boxes"
    />
  </div>
</DocumentFragment>
`;
